### PR TITLE
[plugin-rest-api] Add ability to execute HTTP requests with binary content

### LINKED ISSUE #1739: Add ability to execute HTTP requests with binary content
As a user I want to be able to execute HTTP requests with binary content.

I need to upload media content (images and video files) and to be able to observe them and their previews:

Screenshot 2021-06-21 at 11 56 49

Screenshot 2021-06-21 at 11 58 02

### PR DESCRIPTION
**Pull request by draker94**
Closes #1739

### CODE REVIEW

File: vividus-plugin-rest-api/src/main/java/org/vividus/bdd/steps/api/HttpRequestSteps.java [hunk 1]
```diff
@@ -68,9 +70,21 @@ public HttpRequestSteps(HttpTestContext httpTestContext, HttpRequestExecutor htt
...
+        }
+        else
+        {
+            String resourceDataFullName = content.getResourceDataName();
+            MultipartEntityBuilder multipartEntityBuilder = MultipartEntityBuilder.create();
+            multipartEntityBuilder.addBinaryBody(resourceDataFullName, content.getBytes(), ContentType.DEFAULT_BINARY,
+                    FilenameUtils.getName(resourceDataFullName));
+            httpTestContext.putRequestEntity(multipartEntityBuilder.build());
+        }
     }
```
Comment (valfirst) [inline vividus-plugin-rest-api/src/main/java/org/vividus/bdd/steps/api/HttpRequestSteps.java:88]: my idea in pseudo-code:
```java
public void request(DataWrapper content)
{
    Object data = content.getRawData();
    HttpEntity requestEntity = data instanceof String ? new StringEntity((String) data, StandardCharsets.UTF_8) : new ByteArrayEntity((byte[]) data);
    httpTestContext.putRequestEntity(requestEntity);
}
```

**Review by valfirst — CHANGES_REQUESTED** [review]

File: vividus-plugin-rest-api/src/main/java/org/vividus/bdd/steps/api/HttpRequestSteps.java (hunk 1 above)
Comment (draker94) [reply vividus-plugin-rest-api/src/main/java/org/vividus/bdd/steps/api/HttpRequestSteps.java:88]: > my idea in pseudo-code:
> 
> ```java
> public void request(DataWrapper content)
> {
>     Object data = content.getRawData();
>     HttpEntity requestEntity = data instanceof String ? new StringEntity((String) data, StandardCharsets.UTF_8) : new ByteArrayEntity((byte[]) data);
>     httpTestContext.putRequestEntity(requestEntity);
> }
> ```

And we don't need any name/filename? Just bytes?

Comment (valfirst) [reply vividus-plugin-rest-api/src/main/java/org/vividus/bdd/steps/api/HttpRequestSteps.java:88]: yes, only bytes

PS: name is needed for multipart requests

File: vividus-plugin-rest-api/src/test/java/org/vividus/bdd/steps/api/HttpRequestStepsTests.java [hunk 2]
```diff
@@ -71,18 +72,30 @@
...
         String content = "{\"storeUserNewPassword\": \"зЛЖЛВеБЛги!\"}";
-        httpRequestSteps.request(content);
+        httpRequestSteps.request(new DataWrapper(content));
         verifyPutRequestEntity(content);
     }
 
+    @Test
+    void testRequestBinaryContent() throws IOException
+    {
+        String content = "{there are some binary content}";
```
Comment (valfirst) [inline vividus-plugin-rest-api/src/test/java/org/vividus/bdd/steps/api/HttpRequestStepsTests.java:90]:
```suggestion
        String content = "{there is some binary content}";
```

File: vividus-plugin-rest-api/src/test/java/org/vividus/bdd/steps/api/HttpRequestStepsTests.java [hunk 3]
```diff
@@ -71,18 +72,30 @@
...
+        byte[] contentInBytes = content.getBytes(StandardCharsets.UTF_8);
+        httpRequestSteps.request(new DataWrapper(contentInBytes));
+        ArgumentCaptor<HttpEntity> captor = ArgumentCaptor.forClass(HttpEntity.class);
+        verify(httpTestContext).putRequestEntity(captor.capture());
+        String result = IOUtils.toString(captor.getValue().getContent(), StandardCharsets.UTF_8);
+        assertThat(result, containsString(content));
```
Comment (valfirst) [inline vividus-plugin-rest-api/src/test/java/org/vividus/bdd/steps/api/HttpRequestStepsTests.java:96]: I think it should be `assertEquals` validation here

**Review by valfirst — CHANGES_REQUESTED** [review]

**Review by valfirst — APPROVED** [review]

File: vividus-plugin-rest-api/src/main/java/org/vividus/bdd/steps/api/HttpRequestSteps.java [hunk 4]
```diff
@@ -68,9 +71,12 @@ public HttpRequestSteps(HttpTestContext httpTestContext, HttpRequestExecutor htt
      * @param content HTTP method request body
      */
     @Given("request body: $content")
-    public void request(String content)
+    public void request(DataWrapper content)
```
Comment (ikalinin1) [inline vividus-plugin-rest-api/src/main/java/org/vividus/bdd/steps/api/HttpRequestSteps.java:74]: please update Javadoc and docs to mention that user is responsible for content-type header provision

**Review by uarlouski — APPROVED** [review]

**Review by uarlouski — APPROVED** [review]

File: docs/modules/plugins/pages/plugin-rest-api.adoc [hunk 5]
```diff
@@ -164,6 +164,31 @@ https://www.ietf.org/rfc/rfc3986.txt[RFC 3986].
 
 == HTTP Steps
 
+=== Set HTTP request body
+
+Sets HTTP request body that will be used while executing request. In case of textual content the default HTTP
+request header with name `Content-Type` and value `text/plain; charset=UTF-8` is set. No HTTP request header is set in
+case of binary content.
```
Comment (ikalinin1) [inline docs/modules/plugins/pages/plugin-rest-api.adoc:171]:
```suggestion
Sets HTTP request body that will be used while executing the request. In the case of textual content the default HTTP
request header with name `Content-Type` and value `text/plain; charset=UTF-8` is set. 

WARNING: No HTTP request header is set in the case of binary content.
```

File: docs/modules/plugins/pages/plugin-rest-api.adoc [hunk 6]
```diff
@@ -164,6 +164,31 @@ https://www.ietf.org/rfc/rfc3986.txt[RFC 3986].
...
+[source,gherkin]
+----
+Given request body: Hello!
+----
+
+.Set binary HTTP request body
+[source,gherkin]
+----
+Given request body: #{loadBinaryResource(/data/image.png)}
+----
```
Comment (ikalinin1) [inline docs/modules/plugins/pages/plugin-rest-api.adoc:190]:
```suggestion
----
Given request body: #{loadBinaryResource(/data/image.png)}
When I set request headers:
|name        |value    |
|Content-Type|image/png|
When I send HTTP POST to the relative URL '/upload/png'
----
```

**Pull request merged**